grl/run.py

```diff
@@ -70,32 +71,59 @@ def run_algos(spec, no_gamma, n_random_policies, use_grad, n_steps, max_rollout_
 
     #     logging.info('\n-----------')
 
+def heatmap(spec, no_gamma, num_ticks=5):
+    mdp = MDP(spec['T'], spec['R'], spec['gamma'])
+    amdp = AbstractMDP(mdp, spec['phi'], p0=spec['p0'])
+    policy_eval = PolicyEval(amdp, no_gamma, verbose=False)
+
+    discrepancies = []
+    x = y = np.linspace(0, 1, num_ticks)
+    for i in range(num_ticks):
+        p = x[i]
+        for j in range(num_ticks):
+            q = y[j]
+            pi = np.array([[p, 1 - p], [q, 1 - q], [0, 0]])
+            discrepancies.append(policy_eval.mse_loss(pi))
+
+            if (num_ticks * i + j + 1) % 10 == 0:
+                print(f'Calculating policy {num_ticks * i + j + 1}/{num_ticks * num_ticks}')
+
+    ax = sns.heatmap(np.array(discrepancies).reshape((num_ticks, num_ticks)),
+                     xticklabels=x.round(3),
+                     yticklabels=y.round(3))
+    ax.invert_yaxis()
+    ax.set(xlabel='2nd obs', ylabel='1st obs')
+    ax.set_title(args.spec)
+    plt.show()
```

Comment:
Can you split this up to make things more abstract?

Like one part takes in a 2D array of heat values (and maybe some labels) and it just makes the plot for you. The other part computes that 2D array.

Reply:
You mean into two functions?

Reply:
Maybe that's overkill, now that I think about it more.

Reply:
Never mind!

Reply:
Agree, it's good for now and we can abstract if we ever need some of the inner parts for something else.